### PR TITLE
8292037: Consider moving Linker::methodType as FunctionDescriptor::toMethodType

### PR DESCRIPTION
**Pull request by JornVernee**
This patch moves `Linker::methodType` to `FunctionDescriptor` as `carrierMethodType`.

I went with the name `carrierMethodType` instead of `toMethodType` (as suggested by the JBS issue) since I think it's a bit more descriptive. Please let me know your thoughts.

This issue also suggest adding support for `SequenceLayout` in the linker implementation. I think adding support in FunctionDescriptor could be a good idea, but since passing arrays by value is not supported in C, I feel like we should reject it in the linker implementations we have (which are all C as well). Of course, we could make it work any way by interpreting such cases as a pointer layout instead, but it's a case of being able to do the same thing in multiple ways (which I think can be confusing), and might actually be surprising if someone passes the wrong layout to the linker by accident, but it doesn't get "caught".

I've added support for `SequenceLayout` to `FunctionDescriptor`, but it is still rejected in the type classification logic.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8292037](https://bugs.openjdk.org/browse/JDK-8292037): Consider moving Linker::methodType as FunctionDescriptor::toMethodType

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-foreign pull/717/head:pull/717` \
`$ git checkout pull/717`

Update a local copy of the PR: \
`$ git checkout pull/717` \
`$ git pull https://git.openjdk.org/panama-foreign pull/717/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 717`

View PR using the GUI difftool: \
`$ git pr show -t 717`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-foreign/pull/717.diff">https://git.openjdk.org/panama-foreign/pull/717.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java [hunk 1]
```diff
@@ -94,6 +95,22 @@ public sealed interface FunctionDescriptor permits FunctionDescriptorImpl {
...
     FunctionDescriptor dropReturnLayout();
 
+    /**
+     * Returns the method type consisting of the carrier types of the layouts in this function descriptor.
+     * <p>
+     * The carrier type of a layout is determined as follows:
+     * <ul>
+     * <li>If the layout is a {@link ValueLayout} the carrier type is determined through {@link ValueLayout#carrier()}.</li>
+     * <li>If the layout is a {@link GroupLayout}, or {@link SequenceLayout} the carrier type is {@link MemorySegment}.</li>
+     * <li>If the layout is a {@link PaddingLayout} an {@link IllegalArgumentException} is thrown.</li>
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java:105]: This comment seems to be inconsistent with the @throws clause

Comment (mcimadamore) [reply src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java:105]: The implementation seems to support sequence layouts, and return MemorySegment. Not super sure about that (given the linker doesn't support that).

**Review by mcimadamore — COMMENTED** [review]
I think for now your choice is ok. Depending on whether we land with GroupLayout we might need to revisit though. Note that a sequence layout is just another way to describe a struct layout - so I don't think (as you seem to suggest) that we should pass by pointer, but we should pass by value instead (expanding the sequence layout to the corresponding struct layout).

So, if GroupLayout only covers structs and unions (as now), what we have now makes sense. If GroupLayout is enhanced to cover all composite layouts (including SequenceLayout), then I think the current position seems a tad more inconsistent, as one might like to be able to use GroupLayout for all things that I can pass by value to the linker.

File: src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java [hunk 2]
```diff
@@ -94,6 +95,22 @@ public sealed interface FunctionDescriptor permits FunctionDescriptorImpl {
...
+     * </ul>
+     *
+     * @return the method type consisting of the carrier types of the layouts in this function descriptor
+     * @throws IllegalArgumentException if one or more layouts in the function descriptor are not supported
+     * (e.g. if they are sequence layouts or padding layouts).
+     */
+    MethodType carrierMethodType();
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java:112]: this name seems long, and the extra wording doesn't seem to add a lot (at least IMHO).

File: src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java (hunk 1 above)
Comment (JornVernee) [reply src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java:105]: In the interest of moving this forward, for now I'll keep the current mapping behavior from Linker.methodType, and just do the move to FunctionDescriptor in this patch.

Comment (mcimadamore) [reply src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java:105]: Note that even if we keep behavior, the javadoc `@throws` seems still inconsistent with what the method javadoc says.

File: src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java [hunk 3]
```diff
@@ -94,6 +95,22 @@ public sealed interface FunctionDescriptor permits FunctionDescriptorImpl {
...
+     * The carrier type of a layout is determined as follows:
+     * <ul>
+     * <li>If the layout is a {@link ValueLayout} the carrier type is determined through {@link ValueLayout#carrier()}.</li>
+     * <li>If the layout is a {@link GroupLayout} the carrier type is {@link MemorySegment}.</li>
+     * <li>If the layout is a {@link PaddingLayout}, or {@link SequenceLayout} an {@link IllegalArgumentException} is thrown.</li>
+     * </ul>
+     *
+     * @return the method type consisting of the carrier types of the layouts in this function descriptor
+     * @throws IllegalArgumentException if one or more layouts in the function descriptor can not be mapped to carrier
+     *                                  types (e.g. if they are sequence layouts or padding layouts).
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java:110]:
```suggestion
     *                                  types (e.g. if they are padding layouts).
```

Comment (JornVernee) [reply src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java:110]: I've removed the support for SequenceLayout now, so I think this is currently correct?

Comment (mcimadamore) [reply src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java:110]: Ah - of course. I didn't notice the bullet being rewritten. I think this is a good move, and one that still give us the option to change things later, if needed (it's always relatively ok to turn throwing cases into non-throwing ones).

**Review by mcimadamore — APPROVED** [review]

**Pull request closed without merging**